src/RedirectMiddleware.php

```diff
@@ -98,7 +99,18 @@ public function checkRedirect(
             );
         }
 
-        return $this($nextRequest, $options);
+        /** @var PromiseInterface|ResponseInterface $response */
+        $options['_previous_uris'][] = $request->getUri();
+        $response = $this($nextRequest, $options);
+
+        // Add headers to be able to track history of redirects.
+        if (!empty($options['allow_redirects']['track_redirects'])) {
+            $response = $response->wait(TRUE);
```

Comment:
We don't want to wait here. Instead, you should return a new promise that modifies the response before it is returned.